cmd/dockerd/daemon.go

```diff
@@ -360,8 +362,10 @@ func (cli *daemonCLI) start(ctx context.Context) (err error) {
 		return errors.Wrap(err, "shutting down due to ServeAPI error")
 	}
 
-	if err := tp.Shutdown(context.Background()); err != nil {
-		log.G(ctx).WithError(err).Error("Failed to shutdown OTEL tracing")
+	if otelShutdown != nil {
+		if err := otelShutdown(context.WithoutCancel(ctx)); err != nil {
+			log.G(ctx).WithError(err).Error("Failed to shutdown OTEL tracing")
+		}
 	}
```

Comment:
Not important; wondering if we should make `newTracerProvider` always return a "closer" function that's already wired up, so that we can just;

```go
tp, otelShutdown := newTracerProvider(ctx)
defer otelShutdown()
```

Reply:
We can't use `defer` here, because that would get executed after the final `Daemon shutdown complete` message 😅 

https://github.com/moby/moby/blob/adbf9a25835e0291937c9730089c26f9be84dc0d/cmd/dockerd/daemon.go#L371

I wouldn't bother with that tbh. Except maybe make the cleanup function a non-nil so the caller doesn't have to bother with a nil check.
Also, storing context in a closure isn't really the best practice, and it should be up to the caller to pass a desired context.

Reply:
Removed the nil cleanup func

Reply:
> We can't use defer here, because that would get executed after the final Daemon shutdown complete message 

Heh, yeah, I thought about that part, but chose to ignore that fact 😇 😉 

> Also, storing context in a closure isn't really the best practice, and it should be up to the caller to pass a desired context.

That's a good point!

> Removed the nil cleanup func

Thanks!